### DISCUSSION TITLE
Reasoning behind 0001-Drop-bc-usage-and-remove-symlink-creation-functions.patch

### DISCUSSION THREAD

**Discussion by andregirard37** [OE4T/meta-tegra#1513] [Q&A]
Hello,
I have problems with my Yocto build targeting the NX Orin. I noticed, looking at the NVIDIA forum, that people had /etc/nvpmodel/nvpmodel_p3767_0000.conf populated. my /etc/nvpmodel/ folder is empty. Which is normal since there is a patch to actually remove it: recipes-bsp/tegra-binaries/tegra-nvpower/0001-Drop-bc-usage-and-remove-symlink-creation-functions.patch 
I was wondering the reasoning behind this as my nvpmodel -q --verbose does not seem to work:
$ sudo nvpmodel -q --verbose
NVPM VERB: Config file: /etc/nvpmodel.conf
NVPM VERB: parsing done for /etc/nvpmodel.conf
NVPM VERB: Using default settings.
NVPM WARN: power mode is not set!
NVPM VERB: Using default settings.
I am using the seeed Studios reComputer J4012 devkit. Not sure if that would play any role.

Also, do you know of any way to avoid a reboot of the board when changing the nvpmodel profile:
ubuntu@jetson:~$ sudo nvpmodel -m 2
NVPM ERROR: Error writing 240 to /sys/devices/gpu.0/tpc_pg_mask: 19
NVPM WARN: Reboot required for changing to this power mode: 2
NVPM WARN: DO YOU WANT TO REBOOT NOW? enter YES/yes to confirm:

On the Xavier NX, we used to be able to change the profile on the fly without a reboot. Not sure if that could still be possible.

Thanks as always!

**Comment by ichergui** [comment]
Hi @andregirard37 
Could you please share the branch name ? 
Are you using the `tegra-demo-distro` ? 
I can see different prompt than the one we provide through `tegra-demo-distro` ? Are you sure that you are using Yocto build. 
**Note:** you don't need to use `sudo` when using `tegra-demo-distro`

**Reply by andregirard37** [reply]
Hello. No I am using a custom distro for a project. I am working off Kirkstone and I saw the patch on meta-tegra

**Comment by madisongh** [comment]
The symlinks aren't done at runtime so that we can support building with read-only root filesystems.  Instead, the `NVPMODEL` and `NVPOWER` variables in the machine configuration are used to choose the configuration files for the `nvpmodel` and `nvpower` services at build time.

Orins do require a reboot to change the `nvpmodel`.

**Reply by andregirard37** [reply]
Thanks. So do you mean that in my build/conf/local.conf I need to add NVPMODEL/NVPOWER variables to set the default power profile ? In my /build/conf/local.conf I can add:

NVPMODEL ?= "nvpmodel_p3767_0000"
NVPOWER ?= "jetsonpower_t234"

Since my MACHINE = "p3509-a02-p3767-0000"

I will try that then. Thanks!

**Reply by madisongh** [reply]
Those settings should already be in the `p3509-a02-p3767-0000.conf` file, unless you have modified it to remove them.

Are you not seeing the `/etc/nvpmodel.conf` file (or the `/etc/nvpower/libjetsonpower/jetsonpower_t234.conf` file) at all on the target device?  Or is one there but has the wrong content?

You should also check to see that the `nvpower` and `nvpmodel` services started without error at boot time.

And just to be sure, this is an Orin NX 16GB module (P3767-0000) you're using, right?

**Reply by andregirard37** [reply]
Well, adding those variables did not change anything so they must already be in p3509-a02-p3767-0000.conf as you suggest. 
I see /etc/nvpmodel.conf and /etc/nvpower/libjetsonpower/jetsonpower_t234.conf correctly.

I see:
root@Orin-NX:~# dmesg | grep nvpmodel
[    4.082124] nvpmodel: initialized successfully
root@Orin-NX:~# dmesg | grep nvpower
root@Orin-NX:~#

So nothing for nvpower on dmesg.

I am using Orin NX 16GB module (P3767-0000), from the devkit from https://www.seeedstudio.com/reComputer-J4012-w-o-power-adapter-p-5628.html?queryID=639ef60cde4a38ccc9ff2f82070d4854&objectID=5628&indexName=bazaar_retailer_products

**Reply by madisongh** [reply]
That all sounds OK, so I'm not sure why you're seeing the "power mode is not set" output on the query.
Are you using systemd or sysvinit?  There are some startup ordering dependencies between the nvpower service and nvpmodel that cropped up with L4T R35.4.1... I see a commit that fixed those dependencies for the systemd case, but I don't see corresponding changes for the initscripts.

**Reply by andregirard37** [reply]
Thanks. I am using systemd.

**Reply by madisongh** [reply]
OK, then check the logs `journalctl -b -u nvpower` and `journalctl -b -u nvpmodel` and see if they're really completing successfully.

**Reply by andregirard37** [reply]
Well, the journalctl is not found, so maybe I am using sysvinit, not sure to be honest. I tried the following:
root@Orin-NX:/dev# /etc/init.d/nvpower restart                                                                                                                                                                   
Running NVIDIA power management setup: [OK]
but I still got the same message for: sudo nvpmodel -q --verbose:
NVPM VERB: Config file: /etc/nvpmodel.conf
NVPM VERB: parsing done for /etc/nvpmodel.conf
NVPM VERB: Using default settings.
NVPM WARN: power mode is not set!
NVPM VERB: Using default settings.

Is there an easy way to tell if I am using systemd or sysvinit ? Thanks

**Reply by madisongh** [reply]
The build-time variables that control this are described in [this section of the Yocto Project manual](https://docs.yoctoproject.org/4.0.16/singleindex.html#selecting-an-initialization-manager).

**Reply by andregirard37** [reply]
Thank you. I don't have INIT_MANAGER defined in my local.conf file, so I would assume sysvinit is used as it is the default one for Poky. In fact I just tried a build using INIT_MANAGER = systemd and there is now 1900+ missing recipes to bake. For debugging purposes, I can try to get this build with systemd working and see if I can have journalctl -b -u nvpower and journalctl -b -u nvpmodel to work

**Reply by andregirard37** [reply]
I created the build using systemd but I don't see any info:
root@Orin-NX:~# journalctl -b -u nvpmodel                                                                                                                                                                        
-- No entries --
root@Orin-NX:~# journalctl -b -u nvpower 
-- No entries --

And again:
root@Orin-NX:~# dmesg | grep nvpmodel                                                                                                                                                                        
[    4.011936] nvpmodel: initialized successfully
root@Orin-NX:~# dmesg | grep nvpower 
root@Orin-NX:~#

However, I tried playing around with it and have:

root@Orin-NX:~# sudo nvpmodel -q --verbose
Failed to read config file path, try default
NVPM VERB: Config file: /etc/nvpmodel.conf
NVPM VERB: parsing done for /etc/nvpmodel.conf
NVPM VERB: Using default settings.
NVPM WARN: **power mode is not set!**
NVPM VERB: Using default settings.


root@Orin-NX:~# sudo nvpmodel -m 2
[  225.734823] CPU4: shutdown
[  225.734917] psci: CPU4 killed (polled 0 ms)
[  225.782632] CPU5: shutdown
[  225.782723] psci: CPU5 killed (polled 0 ms)
[  225.822560] CPU6: shutdown
[  225.822667] psci: CPU6 killed (polled 0 ms)
[  225.846838] IRQ 114: no longer affine to CPU7
[  225.847491] CPU7: shutdown
[  225.847582] psci: CPU7 killed (polled 0 ms)
[  225.991757] nvgpu: 17000000.ga10b                  gk20a_scale_init:541  [INFO]  enabled scaling for GPU
[  225.991757] 

root@Orin-NX:~# sudo nvpmodel -m 1
[  257.643620] nvgpu: 17000000.ga10b                 fbp_pg_mask_store:981  [INFO]  no value change, same mask already set
[  257.643956] nvgpu: 17000000.ga10b                 tpc_pg_mask_store:1070 [INFO]  no value change, same mask already set

root@Orin-NX:~# sudo nvpmodel -q --verbose                                                                                                                                                         
NVPM VERB: Config file: /etc/nvpmodel.conf
NVPM VERB: parsing done for /etc/nvpmodel.conf
NVPM VERB: Current mode: **NV Power Mode: 10W
1**
NVPM VERB: PARAM CPU_ONLINE: ARG CORE_0: PATH /sys/devices/system/cpu/cpu0/online: REAL_VAL: 1 CONF_VAL: 1
NVPM VERB: PARAM CPU_ONLINE: ARG CORE_1: PATH /sys/devices/system/cpu/cpu1/online: REAL_VAL: 1 CONF_VAL: 1
NVPM VERB: PARAM CPU_ONLINE: ARG CORE_2: PATH /sys/devices/system/cpu/cpu2/online: REAL_VAL: 1 CONF_VAL: 1
NVPM VERB: PARAM CPU_ONLINE: ARG CORE_3: PATH /sys/devices/system/cpu/cpu3/online: REAL_VAL: 1 CONF_VAL: 1
NVPM VERB: PARAM CPU_ONLINE: ARG CORE_4: PATH /sys/devices/system/cpu/cpu4/online: REAL_VAL: 0 CONF_VAL: 0
NVPM VERB: PARAM CPU_ONLINE: ARG CORE_5: PATH /sys/devices/system/cpu/cpu5/online: REAL_VAL: 0 CONF_VAL: 0
NVPM VERB: PARAM CPU_ONLINE: ARG CORE_6: PATH /sys/devices/system/cpu/cpu6/online: REAL_VAL: 0 CONF_VAL: 0
NVPM VERB: PARAM CPU_ONLINE: ARG CORE_7: PATH /sys/devices/system/cpu/cpu7/online: REAL_VAL: 0 CONF_VAL: 0
NVPM VERB: PARAM FBP_POWER_GATING: ARG FBP_PG_MASK: PATH /sys/devices/gpu.0/fbp_pg_mask: REAL_VAL: 2 CONF_VAL: 2
NVPM VERB: PARAM TPC_POWER_GATING: ARG TPC_PG_MASK: PATH /sys/devices/gpu.0/tpc_pg_mask: REAL_VAL: 252 CONF_VAL: 252
NVPM VERB: PARAM GPU_POWER_CONTROL_ENABLE: ARG GPU_PWR_CNTL_EN: PATH /sys/devices/gpu.0/power/control: REAL_VAL: auto CONF_VAL: on
NVPM VERB: PARAM CPU_A78_0: ARG MIN_FREQ: PATH /sys/devices/system/cpu/cpu0/cpufreq/scaling_min_freq: REAL_VAL: 729600 CONF_VAL: 729600
NVPM VERB: PARAM CPU_A78_0: ARG MAX_FREQ: PATH /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq: REAL_VAL: 1190400 CONF_VAL: 1190400
NVPM VERB: PARAM GPU: ARG MIN_FREQ: PATH /sys/devices/17000000.ga10b/devfreq_dev/min_freq: REAL_VAL: 306000000 CONF_VAL: 0
NVPM VERB: PARAM GPU: ARG MAX_FREQ: PATH /sys/devices/17000000.ga10b/devfreq_dev/max_freq: REAL_VAL: 612000000 CONF_VAL: 612000000
NVPM VERB: PARAM GPU_POWER_CONTROL_DISABLE: ARG GPU_PWR_CNTL_DIS: PATH /sys/devices/gpu.0/power/control: REAL_VAL: auto CONF_VAL: auto
NVPM VERB: PARAM EMC: ARG MAX_FREQ: PATH /sys/kernel/nvpmodel_emc_cap/emc_iso_cap: REAL_VAL: 2133000000 CONF_VAL: 2133000000
NVPM VERB: PARAM DLA0_CORE: ARG MAX_FREQ: PATH /sys/devices/platform/13e40000.host1x/15880000.nvdla0/acm/clk_cap/dla0_core: REAL_VAL: 153600000 CONF_VAL: 153600000
NVPM VERB: PARAM DLA1_CORE: ARG MAX_FREQ: PATH /sys/devices/platform/13e40000.host1x/158c0000.nvdla1/acm/clk_cap/dla1_core: REAL_VAL: 115200000 CONF_VAL: 115000000
NVPM VERB: PARAM DLA0_FALCON: ARG MAX_FREQ: PATH /sys/devices/platform/13e40000.host1x/15880000.nvdla0/acm/clk_cap/dla0_falcon: REAL_VAL: 115200000 CONF_VAL: 115000000
NVPM VERB: PARAM DLA1_FALCON: ARG MAX_FREQ: PATH /sys/devices/platform/13e40000.host1x/158c0000.nvdla1/acm/clk_cap/dla1_falcon: REAL_VAL: 115200000 CONF_VAL: 115000000
NVPM VERB: PARAM PVA0_VPS: ARG MAX_FREQ: PATH /sys/devices/platform/13e40000.host1x/16000000.pva0/acm/clk_cap/pva0_vps: REAL_VAL: 115200000 CONF_VAL: 115000000
NVPM VERB: PARAM PVA0_AXI: ARG MAX_FREQ: PATH /sys/devices/platform/13e40000.host1x/16000000.pva0/acm/clk_cap/pva0_cpu_axi: REAL_VAL: 115200000 CONF_VAL: 115000000

So it seems that I need to force it the first time.

**Reply by madisongh** [reply]
> So it seems that I need to force it the first time.

That could be because the service didn't actually run at boot time.  What do `systemctl status nvpower` and `systemctl status nvpmodel` show?  They should both be "active", IIRC

**Reply by andregirard37** [reply]
Looks like they are inactive:
sh-5.1# systemctl status nvpower
* nvpower.service - NVIDIA power management setup
     Loaded: loaded (8;;file://Orin-NX/lib/systemd/system/nvpower.service/lib/systemd/system/nvpower.service8;;; enabled; vendor prese8;;>
     Active: inactive (dead)

sh-5.1# systemctl status nvpmodel                                                                                                                                                                                  
* nvpmodel.service - NVIDIA power model daemon
     Loaded: loaded (8;;file://Orin-NX/lib/systemd/system/nvpmodel.service/lib/systemd/system/nvpmodel.service8;;; enabled; vendor pres8;;>
     Active: inactive (dead)

**Reply by madisongh** [reply]
Sounds like those services didn't run at boot time, so something's not quite right with the overall systemd setup
